### PR TITLE
Add section about LSM programs

### PR DESCRIPTION
**Pull request by vadorovsky**
Documentation describing LSM and development of LSM eBPF programs. It
contains an example of LSM program which denies setting a nice value for
the userspace process loading it.

Signed-off-by: Michal Rostecki <vadorovsky@gmail.com>

### CODE REVIEW

File: src/programs/lsm.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+## What is LSM
+
+LSM stands for [Linux Security Modules](https://en.wikipedia.org/wiki/Linux_Security_Modules)
+which is a framework which allows developers to write security systems on top
+of the Linux kernel. It's also briefly described in
+[the Linux kernel documentation](https://www.kernel.org/doc/html/latest/security/lsm.html).
+
+LSM is used by kernel modules or (since kernel 5.7) by eBPF programs. The most
+popular modules using LSM are AppArmor, SELinux,
```
Comment (mstange) [inline src/programs/lsm.md:11]: This sentence seems unfinished.

Comment (vadorovsky) [reply src/programs/lsm.md]: Whoops, sorry, fixed

File: src/programs/lsm.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+header inside the Linux kernel. Each LSM hook usually has some arguments with
+information about each action which should make it easier for modules/programs
+to make a decision, which are listed in the
+[lsm_hook_defs.h](https://github.com/torvalds/linux/blob/master/include/linux/lsm_hook_defs.h)
+header.
+
+Modules or eBPF programs based on LSM are triggered by those hooks and then
+they can decide whether a particular action is allowed to happen or not.
+
+For example, there is an LSM hook named `task_setnice` with the following
```
Comment (alessandrod) [inline src/programs/lsm.md:44]: For example consider the `task_setnice` hook, which has the following definition:

File: src/programs/lsm.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+definition:
+
+```c
+LSM_HOOK(int, 0, task_setnice, struct task_struct *p, int nice)
+```
+
+It's getting triggered when a nice value is set for any process in the system.
```
Comment (alessandrod) [inline src/programs/lsm.md:51]: The hook is triggered

File: src/programs/lsm.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+If you are not familiar with the "nice value" concept, check out
```
Comment (alessandrod) [inline src/programs/lsm.md:52]: If you are not familiar with the concept of process niceness, ...

File: src/programs/lsm.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+[this article](https://en.wikipedia.org/wiki/Nice_(Unix)). This LSM hook comes
```
Comment (alessandrod) [inline src/programs/lsm.md:53]: As you can see from the definition, this hook takes the following arguments:

File: src/programs/lsm.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+with the following arguments:
+
+* `p` is the instance of `task_struct` which represents the process on which
+  the nice value is set
+* `nice` is the nice value
+
+Developers of module/program attaching to that hook can decide whether
```
Comment (alessandrod) [inline src/programs/lsm.md:60]: By attaching to the hook, an eBPF program can decide whether to accept or reject the given `nice` value.

File: src/programs/lsm.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+particular actions of setting a nice value to a process are allowed or not.
+
+eBPF LSM programs have access to all those arguments and also one additional
```
Comment (alessandrod) [inline src/programs/lsm.md:63]: In addition to the arguments found in the hook definition, eBPF programs have access to one extra argument ...

File: src/programs/lsm.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+argument - `ret` - which is a return value of potential previous eBPF LSM
+programs.
+
+## Writing LSM BPF program
+
+Let's try to create an LSM eBPF program which which is triggered by
+`task_struct` hook. The purpose of this program will be denying to change a
```
Comment (alessandrod) [inline src/programs/lsm.md:70]: you mean by the `task_setnice` hook.

The purpose of the program will be denying setting the nice value lower than 0 (which means higher priority), for a particular process.

The `renice` tool can be used to change niceness values:

```
renice [value] -p [pid]
```

With our eBPF program, we want to make it impossible to call `renice` for a given `pid` with a negative `[value]`.

File: src/programs/lsm.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+```
+
+That command should create a new Aya project with an empty program attaching to
+the `task_setnice` hook. Let's go to its directory:
+
+```bash
+cd lsm-mount
+```
+
+`task_setnice` has one argument which is an instance of
```
Comment (alessandrod) [inline src/programs/lsm.md:96]: One of the arguments passed to the `task_setnice` hook is a pointer to a `task_struct` type [link].

File: src/programs/lsm.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+```
+
+Now it's time to modify the `lsm-nice-ebpf` project and write an actual program
+there. The full program code should look like:
+
+```rust,ignore
+{{#rustdoc_include ../../examples/lsm-nice/lsm-nice-ebpf/src/main.rs:all}}
+```
+
+Let's look at the code in details.
```
Comment (alessandrod) [inline src/programs/lsm.md:115]: in detail

File: src/programs/lsm.md [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+At the beginning, we include the autogenerated binding to `task_struct`:
+
+```rust,ignore
+{{#rustdoc_include ../../examples/lsm-nice/lsm-nice-ebpf/src/main.rs:vmlinux}}
+```
+
+Then we define a global variable `PID`. Here we set its value to `0`, but later
+the loader is going to set it to the actual PID.
```
Comment (alessandrod) [inline src/programs/lsm.md:124]: We initialize the value to `0`, but at runtime the userspace side will patch the value with the actual `pid` we're interested in. [link to set_global API]

File: src/programs/lsm.md [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+{{#rustdoc_include ../../examples/lsm-nice/lsm-nice/src/main.rs:pid}}
+```
+
+And then we set the global variable:
+
+```rust,ignore
+{{#rustdoc_include ../../examples/lsm-nice/lsm-nice/src/main.rs:load}}
+```
+
+After that, we can build and run out project with:
```
Comment (alessandrod) [inline src/programs/lsm.md:161]: our project

**Review by alessandrod — CHANGES_REQUESTED** [review]
This looks awesome! Just some nits (in addition to the submodules stuff discussed on discord)

File: src/programs/lsm.md (hunk 11 above)
Comment (vadorovsky) [reply src/programs/lsm.md:124]: set_global is not in any released aya version yet, so there is no API documentation (on docs.rs)

**Comment by vadorovsky** [comment]
@alessandrod Addressed your nits. Last Discord discussion convinced you against submodules, so I didn't change anything in the example crate.

**Pull request merged**